Core/MaterialExtension.cpp

```diff
@@ -16,66 +16,103 @@
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given oPs probability not in the range [0, 100] %"); 
+  else 
+    Material->Add_oPsComponent(lifetime, probability);
 }
 
-void MaterialExtension::SetoPsFraction(G4double x)
+void MaterialExtension::Add_directComponent( G4double lifetime, G4double probability )
 {
-  if (x < 0 || x > 1) {
-    G4Exception(
-      "MaterialExtension", "ME01", JustWarning, "3gamma fraction can not extend 1; nothing happend"
-    );
-  } else {
-    foPsFraction = x;
-  };
+  if(lifetime < 0. * ns || lifetime > MaterialParameters::directTauMax) 
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given direct lifetime not in the range [0, 0.6] ns"); 
+  else if( probability < 0. || probability > 100.)
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given direct probability not in the range [0, 100] %"); 
+  else 
+    Material->Add_DirectComponent(lifetime, probability);
 }
 
-void MaterialExtension::SetoPsLifetime(G4double x)
+void MaterialExtension::Set_pPsComponent(G4double lifetime, G4double fraction)
 {
-  if (x < 0 || x > MaterialParameters::oPsTauVaccum) {
-    G4Exception(
-      "MaterialExtension", "ME02", JustWarning, "oPs lifetime can not excide 142*ns"
-    );
-  } else {
-    foPslifetime = x;
-  };
+  if( lifetime < 0. * ns || lifetime > 0.2 ) 
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given pPs lifetime not in the range [0, 0.2] ns"); 
+  else if( fraction < 0. || fraction > 1.)
+    G4Exception("MaterialExtension", "ME01", JustWarning, "given pPs fraction not in the range [0, 1]"); 
+  else 
+    Material->Set_pPsComponent(lifetime, fraction);
 }
 
-const std::vector<G4double> MaterialExtension::GetEventsFraction()
+const G4double MaterialExtension::GetLifetime( double randNumber, G4String channel )
 {
-  G4double direct3g = MaterialParameters::fractionDirect3g;
+  if( channel == "ortho2G" )
+    return Material->GetoPs2GLifetimeFromVector(randNumber);
+  else if( channel == "ortho3G" )
+    return Material->GetoPs3GLifetimeFromVector(randNumber);  
+  else if( channel == "para2G" )
+    return Material->GetpPsLifetime();
+  else
+    return Material->GetDirectLifetimeFromVector(randNumber);
+}
 
-  if (foPsFraction == 1) {
-    direct3g = 0.0f;
-  }
+void MaterialExtension::ChangeMaterialConstants()
+{
+  Material->SetMaterialByName("");
+}
 
-  if ( fPickOffFraction == 1) {
-    direct3g = 0.0f;
+const std::vector<G4double> MaterialExtension::GetEventsFraction()
+{
+  std::vector<G4double> frac;
+  if(MaterialParameters::AnnihlationMode != "")
+  {
+    if(MaterialParameters::AnnihlationMode == "oPs3G")
+    {
+      frac = {0.,0.,0.,0.,1.};
+      return frac;
+    }
+    else if(MaterialParameters::AnnihlationMode == "oPs2G")
+    {
+      frac = {0.,0.,1.,0.,0.};
+      return frac;        
+    }
+    else if(MaterialParameters::AnnihlationMode == "pPs2G")
+    {
+      frac = {1.,0.,0.,0.,0.};
+      return frac;              
+    }
```

Comment:
I think it  is error. You should return something at the end even if no Annihilation mode is set.
Maybe write Error and return empty frac.